### PR TITLE
added custom exception handling

### PR DESCRIPTION
**Pull request by k-boikov**
### Background
An improved version of https://github.com/Significant-Gravitas/Auto-GPT/pull/3810
Credit to @standbyme for the finding/first PR.

### Changes
Added custom exception handler. Added custom exceptions.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [ ] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes. 
- [x] I have run `black .` and `isort .` against my code to ensure it passes our linter.

### CODE REVIEW

File: autogpt/exceptions/exception_handler.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import sys
+
+from autogpt.exceptions import HandledException
+
+
+def handle_exception(exc_type, exc_value, exc_traceback):
+    """A custom exception handler that manages HandledException"""
+    if issubclass(exc_type, HandledException):
+        sys.stderr.write(f"{exc_value.message}\n")
+        sys.exit(1)
```
Comment (Wladastic) [inline autogpt/exceptions/exception_handler.py]: Why would it exit on a handled exception?
I think the original issue was about leaving it up to the user to continue?

Comment (k-boikov) [reply autogpt/exceptions/exception_handler.py]: True, removed.

**Review by lc0rp — DISMISSED** [review]
Do you think we should credit the #3810  author here and close #3810 

Approving to kick-off workflow :)

**Comment by lc0rp** [comment]
@k-boikov let's please add [standbyme](https://github.com/standbyme) 
Ref: https://github.com/Significant-Gravitas/Auto-GPT/pull/3810#issuecomment-1554569536

File: autogpt/exceptions/exception_handler.py (hunk 1 above)
Comment (richbeales) [reply autogpt/exceptions/exception_handler.py]: sys.exit(1) is still there?

Comment (k-boikov) [reply autogpt/exceptions/exception_handler.py]: Uhm...no? I don't see it. The snapshot says it is "outdated".

File: autogpt/setup.py [hunk 2]
```diff
@@ -62,6 +64,8 @@ def prompt_user() -> AIConfig:
     else:
         try:
             return generate_aiconfig_automatic(user_desire)
+        except CriticalException:
```
Comment (richbeales) [inline autogpt/setup.py:67]: We should write something to the log here?

Comment (k-boikov) [reply autogpt/setup.py:67]: I dont think so. We are throwing our own custom exception -> we know exactly what happened. Potentially might have it in the global exception handler, integrated with Sentry.

**Comment by eyalk11** [comment]
I really think that you can handle critical /different kind of exceptions in a unified way using function decorators as my approach here: 
https://github.com/Significant-Gravitas/Auto-GPT/pull/4862/commits/3de5e9687d505397bcb0db329fa968c80d58a776

(both can actually be combined) 


---

sending my comment:

In the recent PR , I introduced simple_exception_handling which is a decorator around functions that automatically prints traceback if there was an exception (and you can instruct it what to do). It also identifies if you run it in IDE(pycharm/vscode) , in that case it lets you see the error.  I found it to be helpful in my other project, and I think it is useful  here. 

```
    @simple_exception_handling(lambda_on_exc=lambda: VoidCommandInstance())
    def generate_instance(self, arguments, agent=None):
        return self.instancecls(self, arguments, agent=agent)
```

**Comment by Pwuts** [comment]
> Users may incorrectly set the key of the LLM service provider. In this case, we should prompt the user as early as possible to avoid invalid communication between the user and the system.

I like this. However, custom exceptions should be defined based on the cases in which they are used, not the way in which they are handled. That is up to the parent context. Auto-GPT uses custom exceptions in the `Agent` since #5077.

**Pull request closed without merging**